### PR TITLE
feature: Use custom key in private_key_jwt towards OAuth2/OIDC IdP

### PR DESCRIPTION
**Pull request by strehle**
Why: Key rotation in UAA should be Independent from scenario.

Currently: UAA stores keys (JWKSet) in jwt.token.policy.keys.
The current signing key is defined with jwt.token.policy.activeKeyId. 
This key is used for JWT creation towards UAA clients (standard token flows), but also with private_key_jwt for the JWT creation towards the IdP.

In case of key rotation this can be less flexible than situation before where secrets were used.  Each IdP had a custom relyingPartySecret, different to the client secret of the oauth2 clients in UAA.

Therefore this PR introduces a sub parameter kid which allows to use a different key for private_key_jwt scenario.
With this the UAA clients can rotate keys without dependency to the configured identity providers in UAA:
The key used in private_key_jwt must be in zone configuration -> jwt.token.policy.keys.

### CODE REVIEW

**Comment by strehle** [comment]
sonar green: 
https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2420

documentation
image

File: server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java [hunk 1]
```diff
@@ -129,21 +135,83 @@ void testGetClientAuthenticationParametersNullJwtClientConfiguration() {
...
+    HashMap customClaims = new HashMap<>();
+    customClaims.put("kid", "myKey");
+    config.setJwtClientAuthentication(customClaims);
+    MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+    // When
+    params = jwtClientAuthentication.getClientAuthenticationParameters(params, config);
+    // Then
+    assertTrue(params.containsKey("client_assertion"));
+    assertTrue(params.containsKey("client_assertion_type"));
+    validateClientAssertionOidcComplaint((String) params.get("client_assertion").get(0));
```
Comment (hsinn0) [inline server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java]: It seems that line 179 is unnecessary as the same call is made in line 180.

File: server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java [hunk 2]
```diff
@@ -129,21 +135,83 @@ void testGetClientAuthenticationParametersNullJwtClientConfiguration() {
...
+    when(keyInfo.keyURL()).thenReturn("http://localhost:8080/uaa/token_key");
     when(keyInfo.getSigner()).thenReturn(signer);
-    when(keyInfo.verifierCertificate()).thenReturn(Optional.of(JwtHelperX5tTest.CERTIFICATE_1));
+    when(keyInfo.verifierCertificate()).thenReturn(x509Certificate != null ? Optional.of(x509Certificate): Optional.of(JwtHelperX5tTest.CERTIFICATE_1));
     when(signer.sign(any())).thenReturn("dummy".getBytes());
   }
 
-  private static void validateClientAssertionOidcComplaint(String clientAssertion) throws ParseException {
-    JWTClaimsSet jwtClaimsSet = JWTParser.parse(clientAssertion).getJWTClaimsSet();
+  private static JWSHeader validateClientAssertionOidcComplaint(String clientAssertion) throws ParseException {
```
Comment (hsinn0) [inline server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java]: I think that the method name should be changed to represent the new implementation.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointDocs.java [hunk 3]
```diff
@@ -559,7 +559,11 @@ void createOidcIdentityProvider() throws Exception {
...
                 fieldWithPath("config.setForwardHeader").optional(false).type(BOOLEAN).description("Only effective if Password Grant enabled. Set X-Forward-For header in Password Grant request to this identity provider."),
-                fieldWithPath("config.jwtClientAuthentication").optional(null).type(OBJECT).description("<small><mark>UAA 76.5.0</mark></small> Only effective if relyingPartySecret is not set or null. Creates private_key_jwt client authentication according to OIDC or OAuth2 (RFC 7523) standard. For OIDC, set true. For OAuth2, define custom elements `iss` and `aud`."),
+                fieldWithPath("config.jwtClientAuthentication").optional(null).type(OBJECT).description("<small><mark>UAA 76.5.0</mark></small> Only effective if relyingPartySecret is not set or null. Creates private_key_jwt client authentication according to OIDC or OAuth2 (RFC 7523) standard. "+
+                    "<br>Please note that you can precise the created JWT for client authentication, e.g. if your IdP follows OAuth2 standard RFC 7523. If sub-parameter is not set, default is used, see" +
```
Comment (hsinn0) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointDocs.java]: > default is used, see

Was this a typo or missing words? "see" what?

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointDocs.java [hunk 4]
```diff
@@ -559,7 +559,11 @@ void createOidcIdentityProvider() throws Exception {
                 GROUP_WHITELIST,
                 fieldWithPath("config.passwordGrantEnabled").optional(false).type(BOOLEAN).description("Enable Resource Owner Password Grant flow for this identity provider."),
                 fieldWithPath("config.setForwardHeader").optional(false).type(BOOLEAN).description("Only effective if Password Grant enabled. Set X-Forward-For header in Password Grant request to this identity provider."),
-                fieldWithPath("config.jwtClientAuthentication").optional(null).type(OBJECT).description("<small><mark>UAA 76.5.0</mark></small> Only effective if relyingPartySecret is not set or null. Creates private_key_jwt client authentication according to OIDC or OAuth2 (RFC 7523) standard. For OIDC, set true. For OAuth2, define custom elements `iss` and `aud`."),
+                fieldWithPath("config.jwtClientAuthentication").optional(null).type(OBJECT).description("<small><mark>UAA 76.5.0</mark></small> Only effective if relyingPartySecret is not set or null. Creates private_key_jwt client authentication according to OIDC or OAuth2 (RFC 7523) standard. "+
```
Comment (hsinn0) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointDocs.java:562]: So no more of this?

> For OIDC, set true.

Then is this breaking change?

**Review by hsinn0 — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java (hunk 1 above)
Comment (strehle) [reply server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java]: done

File: server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java (hunk 2 above)
Comment (strehle) [reply server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java]: Refactored this method can therefore stay as it is

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointDocs.java (hunk 4 above)
Comment (strehle) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointDocs.java:562]: yes OIDC and simple true still is valid and possible

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointDocs.java (hunk 3 above)
Comment (strehle) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/providers/IdentityProviderEndpointDocs.java]: changed texts

**Comment by strehle** [comment]
@hsinn0  Thanks for the review. I changed the text. 
This PR is no breaking or any change, it is an enhancement to existing only, hope that gets clear from text now

**Comment by strehle** [comment]
Documentation Update
image

**Review by hsinn0 — APPROVED** [review]

**Pull request merged**